File: patch/scan.js
'use strict';

// Finds balanced bracket spans inside minified JS.
//
// This file knows nothing about what JS means. It only counts pairs without being
// fooled by brackets inside string, template, or regex literals — which is exactly
// why it can be tested apart from the patching logic.
//
// Mirrors src/patch/scan.rs; the two are meant to be read side by side.

/** Index just past the closing quote of the string starting at `i`. */
function skipQuoted(src, i, quote) {
  i += 1;
  while (i < src.length) {
    if (src.charCodeAt(i) === 92) { i += 2; continue; }   // backslash escape
    if (src[i] === quote) return i;
    i += 1;
  }
  return i;
}

/** The same for a template literal, stepping over `${…}` substitutions. */
function skipTemplate(src, i) {
  i += 1;
  while (i < src.length) {
    if (src.charCodeAt(i) === 92) { i += 2; continue; }
    if (src[i] === '`') return i;
    if (src[i] === '$' && src[i + 1] === '{') {
      const close = scanBalanced(src, i + 1);
      if (close < 0) return src.length;
      i = close;
      continue;
    }
    i += 1;
  }
  return i;
}

/** Given `src[open] === '{'`, the index just past its matching `}`, or -1. */
function scanBalanced(src, open) {
  let depth = 0;
  for (let i = open; i < src.length; i += 1) {
    const c = src[i];
    if (c === '"' || c === "'") { i = skipQuoted(src, i, c); continue; }
    if (c === '`') { i = skipTemplate(src, i); continue; }
    if (c === '{') depth += 1;
    else if (c === '}') {
      depth -= 1;
      if (depth === 0) return i + 1;
    }
  }
  return -1;
}

/** The same for parentheses: given `src[open] === '('`, the index of its `)`. */
function scanParens(src, open) {
  let depth = 0;
  for (let i = open; i < src.length; i += 1) {
    const c = src[i];
    if (c === '"' || c === "'") { i = skipQuoted(src, i, c); continue; }
    if (c === '`') { i = skipTemplate(src, i); continue; }
    if (c === '(') depth += 1;
    else if (c === ')') { depth -= 1; if (depth === 0) return i; }
  }
  return -1;
}

/**
 * Locate the object literal that follows `id={`.
 * Checks the preceding character so we do not match the tail of a longer identifier
 * — `ENe` must not be found inside `xENe`.
 */
function findObjectLiteral(src, id) {
  const needle = `${id}={`;
  let from = 0;
  for (;;) {
    const start = src.indexOf(needle, from);
    if (start < 0) return null;

    const prev = start > 0 ? src[start - 1] : '';
    if (/[A-Za-z0-9_$]/.test(prev)) { from = start + 1; continue; }

    const braceStart = start + id.length + 1;
    const end = scanBalanced(src, braceStart);
    if (end < 0) { from = start + 1; continue; }
    return { start: braceStart, end, body: src.slice(braceStart, end) };
  }
}

module.exports = { scanBalanced, scanParens, skipQuoted, skipTemplate, findObjectLiteral };

File: src/patch/scan.rs
//! Finds balanced bracket spans inside minified JS.
//!
//! This file knows nothing about what JS means. It only counts pairs without being
//! fooled by brackets inside string, template, or regex literals — which is exactly
//! why it can be tested apart from the patching logic.

/// Index just past the bracket that closes the one at `open`, or `None` if unbalanced.
pub fn balanced(src: &str, open: usize) -> Option<usize> {
    let b = src.as_bytes();
    let (o, c) = match *b.get(open)? {
        b'{' => (b'{', b'}'),
        b'[' => (b'[', b']'),
        b'(' => (b'(', b')'),
        _ => return None,
    };

    let mut depth = 0i32;
    let mut i = open;
    let mut quote: Option<u8> = None;
    let mut escaped = false;

    while i < b.len() {
        let ch = b[i];
        if let Some(q) = quote {
            if escaped {
                escaped = false;
            } else if ch == b'\\' {
                escaped = true;
            } else if ch == q {
                quote = None;
            }
            i += 1;
            continue;
        }
        match ch {
            b'"' | b'\'' | b'`' => quote = Some(ch),
            x if x == o => depth += 1,
            x if x == c => {
                depth -= 1;
                if depth == 0 {
                    return Some(i + 1);
                }
            }
            _ => {}
        }
        i += 1;
    }
    None
}

/// Locate the object literal that follows `name=`.
/// Checks the preceding byte so we do not match the tail of a longer identifier.
pub fn object_literal(src: &str, name: &str) -> Option<(usize, usize)> {
    let pat = format!("{name}=");
    let mut from = 0;
    while let Some(rel) = src[from..].find(&pat) {
        let at = from + rel;
        let prev_ok = at == 0
            || !src.as_bytes()[at - 1].is_ascii_alphanumeric()
                && src.as_bytes()[at - 1] != b'_'
                && src.as_bytes()[at - 1] != b'$';
        let brace = at + pat.len();
        if prev_ok && src.as_bytes().get(brace) == Some(&b'{') {
            if let Some(end) = balanced(src, brace) {
                return Some((brace, end));
            }
        }
        from = at + pat.len();
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn counts_nested_objects_to_the_end() {
        let s = "X={a:{b:{c:1}},d:2};";
        let (o, e) = object_literal(s, "X").unwrap();
        assert_eq!(&s[o..e], "{a:{b:{c:1}},d:2}");
    }

    #[test]
    fn is_not_fooled_by_a_closing_brace_inside_a_string() {
        let s = r#"X={a:"}}}",b:1};"#;
        let (o, e) = object_literal(s, "X").unwrap();
        assert_eq!(&s[o..e], r#"{a:"}}}",b:1}"#);
    }

    #[test]
    fn steps_over_an_escaped_quote() {
        let s = r#"X={a:"a\"}",b:1};"#;
        let (o, e) = object_literal(s, "X").unwrap();
        assert_eq!(&s[o..e], r#"{a:"a\"}",b:1}"#);
    }

    #[test]
    fn does_not_match_the_tail_of_a_longer_identifier() {
        let s = "xENe={wrong:1};ENe={right:1};";
        let (o, e) = object_literal(s, "ENe").unwrap();
        assert_eq!(&s[o..e], "{right:1}");
    }

    #[test]
    fn unbalanced_input_gives_none() {
        assert_eq!(balanced("{a:1", 0), None);
    }

    #[test]
    fn ignores_braces_inside_a_template_literal() {
        let s = "X={a:`${1+1}}`,b:2};";
        let (o, e) = object_literal(s, "X").unwrap();
        assert_eq!(&s[o..e], "{a:`${1+1}}`,b:2}");
    }
}
